src/bson_mapper/bson_archiver.hpp

```diff
@@ -0,0 +1,1078 @@
+                }
+            }
+
+            // Elements in arrays do not have names.
+            if (topType == OutputNodeType::InArray) return;
+        }
+
+        if (!_nextName && _nodeTypeStack.empty() && nodeInRoot) {
+            // This is a document at the root node.
+            // Do nothing since no name is expected.
+            return;
+        } else if (!_nextName) {
+            // Enforce the existence of a name-value pair if this is not a node in root,
+            // or this in an element in root.
+            throw cereal::Exception("Missing a name for current node or element.\n");
+        } else {
+            // Set the key of this element to the name stored by the archiver.
+            _bsonBuilder.key_owned(_nextName);
+            _nextName = nullptr;
+        }
+    }
+
+    /**
+     * Designates that the current node should be output as an array, not an
+     * object.
+     */
+    void makeArray() {
+        _nodeTypeStack.top() = OutputNodeType::StartArray;
+    }
+
+   private:
+    // The BSONCXX builder for this archive.
+    BSONBuilder _bsonBuilder;
+
+    // The stream to which to write the BSON archive.
+    std::ostream& _writeStream;
+
+    // The name of the next element to be added to the archive.
+    char const* _nextName;
+
+    // A stack maintaining the state of the nodes currently being written.
+    std::stack<OutputNodeType> _nodeTypeStack;
+
+};  // BSONOutputArchive
+
+class BSONInputArchive : public InputArchive<BSONInputArchive>, public traits::TextArchive {
+    typedef std::vector<char> buffer_type;
+    enum class InputNodeType { InObject, InEmbeddedObject, InEmbeddedArray, InRootElement };
+
+   public:
+    /**
+    * Construct a BSONInputArchive from an input stream of BSON data.
+    *
+    * @param stream
+    *    The stream from which to read BSON data.
+    */
+    BSONInputArchive(std::istream& stream)
+        : InputArchive<BSONInputArchive>(this),
+          _nextName(nullptr),
+          _readStream(stream),
+          _readFirstDoc(false),
+          _curBsonDoc(nullptr, 0, [](uint8_t*) {}) {
+    }
+
+   private:
+    /**
+     * Reads the next BSON document from the istream. This should be called whenever
+     * we are starting to load in a root element or root node.
+     */
+    void readNextDoc() {
+        // Determine the size of the BSON document in bytes.
+        // TODO: Only works on little endian.
+        int32_t docsize;
+        char docsize_buf[sizeof(docsize)];
+        _readStream.read(docsize_buf, sizeof(docsize));
+        std::memcpy(&docsize, docsize_buf, sizeof(docsize));
+
+        // Throw an exception if the end of the stream is prematurely reached.
+        if (_readStream.eof() || !_readStream) {
+            throw cereal::Exception("No more data in BSONInputArchive stream.\n");
+        }
+
+        // Create a buffer to store the BSON document.
+        char* bsonData = new char[docsize];
```

Comment:
Prefer:

auto bsonData = bsoncxx::stdx::make_unique<char[]>(docsize);

Or just regular std::make_unique if we want to go to c++14